File: client/ng-mes-auto/src/app/models/package-box.ts
import {Batch} from './batch';
import {Grade} from './grade';
import {Operator} from './operator';
import {PackageClass} from './package-class';
import {SapT001l} from './sapT001l';
import {Workshop} from './workshop';

export const PACKAGE_BOX_TYPE = ['AUTO', 'MANUAL', 'MANUAL_APPEND', 'BIG_SILK_CAR', 'SMALL'];

export class PackageBox {
  id: string;
  type: 'AUTO' | 'MANUAL' | 'MANUAL_APPEND' | 'BIG_SILK_CAR' | 'SMALL';
  code: string;
  batch: Batch;
  grade: Grade;
  saleType: 'DOMESTIC' | 'FOREIGN';
  sapT001l: SapT001l;
  palletCode: string;
  palletType: string;
  packageType: string;
  foamType: string;
  foamNum: number;
  silkCount: number;
  netWeight: number;
  grossWeight: number;
  printCount: number;
  budat: Date;
  budatClass: PackageClass;
  creator: Operator;
  createDateTime: Operator;

  static assign(...sources: any[]): PackageBox {
    const result = Object.assign(new PackageBox(), ...sources);
    return result;
  }

  static toEntities(os: PackageBox[], entities?: { [id: string]: PackageBox }): { [id: string]: PackageBox } {
    return (os || []).reduce((acc, cur) => {
      acc[cur.id] = PackageBox.assign(cur);
      return acc;
    }, {...(entities || {})});
  }

  static isValidPalletCode(palletCode: string): boolean {
    if (!palletCode) {
      return false;
    }
    const check = palletCode.toUpperCase();
    if (check.startsWith('E') || check.startsWith('F')) {
      return true;
    }
    return false;
  }

  static weightFormula(options: { batch: Batch, grade: Grade, silkCount: number, grossWeight?: number, tare?: number }): { netWeight: string; grossWeight?: string } {
    const {batch, grade, silkCount, grossWeight, tare} = options;
    if (batch && grade && silkCount) {
      if (grade.sortBy >= 100) {
        return {
          netWeight: `${silkCount} * ${batch.silkWeight}`,
          grossWeight: `${silkCount} * ${batch.silkWeight} + ${silkCount} * ${batch.tubeWeight}`,
        };
      } else if (grossWeight) {
        let netWeightFormula = `${grossWeight} - ${silkCount} * ${batch.tubeWeight}`;
        if (tare) {
          netWeightFormula = `${netWeightFormula} - ${tare}`;
        }
        if (grade.sortBy >= 80) {
          return {
            netWeight: `${netWeightFormula} - 0.5`,
          };
        } else {
          return {
            netWeight: `${netWeightFormula}`,
          };
        }
      }
    }
    return null;
  }

  static sapT001ls(options: { saleType: string; workshop: Workshop; palletCode?: string; }): SapT001l[] {
    const {saleType, workshop: {sapT001ls, sapT001lsForeign, sapT001lsPallet}, palletCode} = options;
    if (PackageBox.isValidPalletCode(palletCode)) {
      return sapT001lsPallet;
    }
    switch (saleType) {
      case  'DOMESTIC': {
        return sapT001ls;
      }
      case 'FOREIGN': {
        return sapT001lsForeign;
      }
    }
    return [];
  }
}

File: mes-root/mes-auto-core/src/main/java/com/hengyi/japp/mes/auto/domain/PackageBox.java
package com.hengyi.japp.mes.auto.domain;

import com.fasterxml.jackson.annotation.JsonGetter;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.databind.JsonNode;
import com.hengyi.japp.mes.auto.domain.data.PackageBoxType;
import com.hengyi.japp.mes.auto.domain.data.SaleType;
import lombok.*;

import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.Id;
import javax.validation.constraints.Min;
import javax.validation.constraints.NotBlank;
import javax.validation.constraints.NotNull;
import java.util.Collection;
import java.util.Date;

import static com.github.ixtf.japp.core.Constant.MAPPER;

/**
 * @author jzb 2018-06-22
 */
@NoArgsConstructor
@ToString(onlyExplicitlyIncluded = true)
@EqualsAndHashCode(onlyExplicitlyIncluded = true)
@Entity
public class PackageBox implements EntityLoggable {
    @ToString.Include
    @EqualsAndHashCode.Include
    @Getter
    @Setter
    @Id
    @NotBlank
    private String id;
    @ToString.Include
    @Getter
    @Setter
    @Column
    @NotNull
    private PackageBoxType type;
    @ToString.Include
    @Getter
    @Setter
    @Column
    @NotBlank
    private String code;
    @Getter
    @Setter
    @Column
    private SaleType saleType;
    /**
     * 久鼎塑托
     */
    @Getter
    @Setter
    @Column
    private String palletCode;
    /**
     * 托盘类型
     */
    @Getter
    @Setter
    @Column
    private String palletType;
    /**
     * 包装类型
     */
    @Getter
    @Setter
    @Column
    private String packageType;
    /**
     * 泡沫类型
     */
    @Getter
    @Setter
    @Column
    private String foamType;
    /**
     * 泡沫数量
     */
    @Getter
    @Setter
    @Column
    private int foamNum;
    @ToString.Include
    @Getter
    @Setter
    @Column
    @NotNull
    private Batch batch;
    @ToString.Include
    @Getter
    @Setter
    @Column
    @NotNull
    private Grade grade;
    @Getter
    @Setter
    @Column
    @Min(1)
    private int silkCount;
    @Getter
    @Setter
    @Column
    @Min(1)
    private double grossWeight;
    @Getter
    @Setter
    @Column
    @Min(1)
    private double netWeight;
    @JsonIgnore
    @Getter
    @Setter
    @Column
    private Collection<Silk> silks;
    @JsonIgnore
    @Getter
    @Setter
    @Column
    private Collection<SilkCarRecord> silkCarRecords;

    @JsonIgnore
    @Getter
    @Setter
    @Column(name = "command")
    private String commandJsonString;
    @Getter
    @Setter
    @Column
    private String riambJobId;

    /**
     * 唛头上的打印日期
     */
    @Getter
    @Setter
    @Column
    private Date printDate;
    @Getter
    @Setter
    @Column
    private PackageClass printClass;
    @Getter
    @Setter
    @Column
    @Min(0)
    private int printCount;
    @Getter
    @Setter
    @Column
    private String automaticPackeLine;
    /**
     * SAP 入库日期
     */
    @Getter
    @Setter
    @Column
    private Date budat;
    @Getter
    @Setter
    @Column
    private PackageClass budatClass;
    @Getter
    @Setter
    @Column
    private SapT001l sapT001l;
    @Getter
    @Setter
    @Column
    private boolean inWarehouse;

    @JsonIgnore
    @Getter
    @Setter
    @Column
    @NotNull
    private Operator creator;
    @JsonIgnore
    @Getter
    @Setter
    @Column(name = "cdt")
    @NotNull
    private Date createDateTime;
    @JsonIgnore
    @Getter
    @Setter
    @Column
    private Operator modifier;
    @JsonIgnore
    @Getter
    @Setter
    @Column(name = "mdt")
    private Date modifyDateTime;
    @JsonIgnore
    @Getter
    @Setter
    @Column
    private boolean deleted;

//    @SneakyThrows
//    @JsonGetter("command")
//    public JsonNode command() {
//        final String commandJsonString = getCommandJsonString();
//        return J.isBlank(commandJsonString) ? null : MAPPER.readTree(commandJsonString);
//    }

    @SneakyThrows
    public void command(JsonNode jsonNode) {
        if (jsonNode != null) {
            setCommandJsonString(MAPPER.writeValueAsString(jsonNode));
        } else {
            setCommandJsonString(null);
        }
    }

    @JsonGetter("creator")
    public Operator _creator_() {
        return getCreator();
    }
}
